src/full_space.jl

```diff
@@ -0,0 +1,224 @@
+
+    l_con::Vector{Float64} = []
+    u_con::Vector{Float64} = []
+    for i in 1:nb
+        if m.blocks[i].meta.ncon > 0
+            l_con = vcat(l_con, m.blocks[i].meta.lcon)
+            u_con = vcat(u_con, m.blocks[i].meta.ucon)
+        end
+    end
+    l_con = vcat(l_con, get_RHSvector(m))
+    u_con = vcat(u_con, get_RHSvector(m))
+
+    l_var::Vector{Float64} = []
+    u_var::Vector{Float64} = []
+    for i in 1:nb
+        l_var = vcat(l_var, m.blocks[i].meta.lvar)
+        u_var = vcat(u_var, m.blocks[i].meta.uvar)
+    end
+
+    meta = NLPModelMeta{T, Vector{T}}(
+        n_var,
+        ncon = sum(m.blocks[i].meta.ncon for i in 1:nb) + 
+        m.problem_size.link_counter,
+        nnzh = sum(m.blocks[i].meta.nnzh for i in 1:nb),
+        nnzj = sum(m.blocks[i].meta.nnzj for i in 1:nb) + 
+        nnz(get_blockmatrix(m)),
+        x0 = zeros(Float64, n_var),
+        lvar = l_var,
+        uvar = u_var,
+        lcon = l_con,
+        ucon = u_con,
+        minimize = true,
+        name = "full_space",
+    )
+return FullSpaceModel(meta, Counters(), m)
+end
+
+FullSpaceModel(m::AbstractBlockNLPModel) = FullSpaceModel(Float64, m)
+
+function NLPModels.obj(nlp::FullSpaceModel, x::AbstractVector)
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nvar x
+    increment!(nlp, :neval_obj)
+    return sum(obj(nlp.blocknlp.blocks[i].problem_block, 
+    x[nlp.blocknlp.blocks[i].var_idx]) for i in 1:nb)
+end
+
+function NLPModels.grad!(nlp::FullSpaceModel, x::AbstractVector, gx::AbstractVector)
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nvar x gx
+    increment!(nlp, :neval_grad)
+    for i in 1:nb
+        gx[nlp.blocknlp.blocks[i].var_idx] = 
+        grad(nlp.blocknlp.blocks[i].problem_block, x[nlp.blocknlp.blocks[i].var_idx])
+    end
+    return gx
+end
+
+function NLPModels.hess_structure!(nlp::FullSpaceModel, rows::AbstractVector{T}, 
+cols::AbstractVector{T}) where {T}
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nnzh rows cols
+    H = spzeros(nlp.meta.nvar, nlp.meta.nvar)
+    for i in 1:nb
+        H[nlp.blocknlp.blocks[i].var_idx, nlp.blocknlp.blocks[i].var_idx] = 
+        sparse(hess_structure(nlp.blocknlp.blocks[i].problem_block)[1], 
+        hess_structure(nlp.blocknlp.blocks[i].problem_block)[2], 
+        ones(nlp.blocknlp.blocks[i].meta.nnzh), 
+        nlp.blocknlp.blocks[i].meta.nvar, nlp.blocknlp.blocks[i].meta.nvar)
+    end
+    rows, cols, dummy_vals = findnz(H)
+    return rows, cols
+end
+
+function NLPModels.hess_coord!(
+nlp::FullSpaceModel,
+x::AbstractVector{T},
+vals::AbstractVector{T};
+obj_weight = one(T),
+) where {T}
+    nb = nlp.blocknlp.problem_size.block_counter
+    @lencheck nlp.meta.nvar x
+    @lencheck nlp.meta.nnzh vals
+    increment!(nlp, :neval_hess)
+
+    H = spzeros(nlp.meta.nvar, nlp.meta.nvar)
+    for i in 1:nb
+    H[nlp.blocknlp.blocks[i].var_idx, nlp.blocknlp.blocks[i].var_idx] = 
```

Comment:
This code is almost exactly the same as in the previous function. Can we define an intermediate function to avoid duplicating code?

Reply:
Done.